### PR TITLE
Homework6

### PR DESCRIPTION
**Pull request by jfdsmit**

### CODE REVIEW

File: course5/homework6/string.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#ifndef _JDS_STRING_H_
```
Comment (maciekgajewski) [inline course5/homework6/string.h]: ```#pragma once``` is better

File: course5/homework6/string.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	void operator=(String const & that);
+	void operator=(String && that);
+	void operator=(char const * source);
+	unsigned length() const;
+	char& operator[](unsigned index);
+	char operator[](unsigned index) const;
+	String operator+(String const & that) const;
+	String& operator+=(String const & that);
+	char const * c_str() const;
+
```
Comment (maciekgajewski) [inline course5/homework6/string.h]: I've realised that operators == and != are very useful in automating the unit test

File: course5/homework6/string.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+private:
+	std::unique_ptr<char[]> mBuffer;
+	unsigned mLength;
+};
+
+}
+
+std::ostream& operator<<(std::ostream& s, jds::String const & value);
```
Comment (maciekgajewski) [inline course5/homework6/string.h]: There is no reason for it to be in global namespace. You should be able to leave in the class' namespace.

File: course5/homework6/string.cc [hunk 4]
```diff
@@ -0,0 +1,115 @@
+#include <memory>
+#include <cstring>
+
+#include "string.h"
+
+namespace jds {
+
+String::String()
+{
+	mLength = 0;
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: Just set the default value in the class definition, and get away with compiler-generated constructor.

File: course5/homework6/string.cc [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+{
+	mLength = that.mLength;
+	mBuffer = std::make_unique<char[]>(mLength + 1);
+	std::memcpy(this->mBuffer.get(), that.mBuffer.get(), that.mLength);
+	mBuffer.get()[mLength] = '\0';
+}
+
+String::String(String && that)
+{
+	mBuffer = std::unique_ptr<char []>(that.mBuffer.release());
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: Ugly! :scream_cat:
Use ```std::move``` instead

File: course5/homework6/string.cc [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	mLength = that.mLength;
+	that.mLength = 0;
+}
+
+String::String(char const * source)
+{
+	mLength = strlen(source);
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: ```std::strlen```

File: course5/homework6/string.cc [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+String::String()
+{
+	mLength = 0;
+}
+
+String::String(String const & that)
+{
+	mLength = that.mLength;
+	mBuffer = std::make_unique<char[]>(mLength + 1);
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: So, copying from an empty string will allocate buffer of 1?

File: course5/homework6/string.cc [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+{
+	mBuffer = std::unique_ptr<char []>(that.mBuffer.release());
+	mLength = that.mLength;
+	that.mLength = 0;
+}
+
+void String::operator=(char const * source)
+{
+	mLength = strlen(source);
+	mBuffer = std::make_unique<char[]>(mLength + 1);
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: Even doing this:
```
s = "";
```
will allocate.

File: course5/homework6/string.cc [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+unsigned String::length() const
+{
+	return mLength;
+}
+
+char& String::operator[](unsigned index)
+{
+	if (index >= mLength)
```
Comment (maciekgajewski) [inline course5/homework6/string.cc:69]: why ```>=```?
What if your string is default-constructed?

File: course5/homework6/string.cc [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+	{
+		throw std::runtime_error("String index out of bounds");
```
Comment (maciekgajewski) [inline course5/homework6/string.cc]: ```std::range_error``` may be better.

File: course5/homework6/main.cc [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+	assert(std::strcmp((assigned_cstr + test_string).c_str(), "abcdef") == 0);
+	pass();
+
+	announce("operator+=");
+	assigned_cstr += test_string;
+	assert(std::strcmp(assigned_cstr.c_str(), "abcdef") == 0);
+	pass();
+
+	return 0;
+}
```
Comment (maciekgajewski) [inline course5/homework6/main.cc]: - tests fro const object are missing
- test that copying is indeed deep (modyfying copy doesn't modify the source)
- test corner cases - assigning from empty, to empty, adding two empty etc

**Review by maciekgajewski — CHANGES_REQUESTED** [review]
Nice and clean code, but:
- unnecessary allocations
- expand tests

File: course5/homework6/string.cc (hunk 7 above)
Comment (jfdsmit) [reply course5/homework6/string.cc]: Yes it would and no it should not do that

File: course5/homework6/string.cc (hunk 9 above)
Comment (jfdsmit) [reply course5/homework6/string.cc:69]: A string with length x has only valid indices 0 through x-1, hence if index == length it's already out of range. If the string is default constructed it's empty and has no accessible characters.

If you don't like >= and <= the logic could be inverted to operate normally when index < mLength and error out in the other branch.

Comment (maciekgajewski) [reply course5/homework6/string.cc:69]: yes, of course, it is correct. My bad

File: course5/homework6/string.h (hunk 2 above)
Comment (jfdsmit) [reply course5/homework6/string.h]: Yes, I stuck too religiously to the listing in the e-mail

File: course5/homework6/string.cc (hunk 5 above)
Comment (jfdsmit) [reply course5/homework6/string.cc]: Tried that first last night and did not get it to compile for some reason so I thought I was doing something wrong. This worked too. Your assertion that `std::move` should work made me put it back in and apparently I did something wrong yesterday because it indeed "just works".

File: course5/homework6/main.cc (hunk 11 above)
Comment (jfdsmit) [reply course5/homework6/main.cc]: I'm not exactly sure what "tests for const object" implies, could you elaborate what should be tested about const objects?

Comment (maciekgajewski) [reply course5/homework6/main.cc]:
```
const String cabc = "abc";
assert(cabc.length() == 3);
assert(cabc[1] = 'b');
assert(cabc == "abc");
// etc..
```
